icu4c/source/common/charstr.cpp

```diff
@@ -197,7 +197,14 @@ CharString &CharString::appendPathPart(StringPiece s, UErrorCode &errorCode) {
     }
     char c;
     if(len>0 && (c=buffer[len-1])!=U_FILE_SEP_CHAR && c!=U_FILE_ALT_SEP_CHAR) {
-        append(U_FILE_SEP_CHAR, errorCode);
+        char dirSepChar = U_FILE_SEP_CHAR;
+#if (U_FILE_SEP_CHAR != U_FILE_ALT_SEP_CHAR)
+        // We may need to append a different directory separator when building for Cygwin or MSYS2.
+        if(strchr(data(), U_FILE_SEP_CHAR) == NULL && strchr(data(), U_FILE_ALT_SEP_CHAR) != NULL) {
```

Comment:
This change is ok in principle.

1. Internally we don't use strchr() and such directly but instead use wrappers with a uprv_ prefix, like uprv_strchr().
2. However, we don't forbid embedded NULs, and we have the length, so we should use uprv_memchr() with pointer+length.
3. Please use nullptr instead of NULL.

Reply:
ad 2. I think, that's then an invalid path, and it doesn't matter if we return U_FILE_SEP_CHAR or U_FILE_ALT_SEP_CHAR.